include/triton/core/tritonserver.h

```diff
@@ -2705,6 +2710,28 @@ TRITONSERVER_DECLSPEC struct TRITONSERVER_Error* TRITONSERVER_MetricIncrement(
 TRITONSERVER_DECLSPEC struct TRITONSERVER_Error* TRITONSERVER_MetricSet(
     struct TRITONSERVER_Metric* metric, double value);
 
+/// Observe the current value of metric to value.
+/// Supports metrics of kind TRITONSERVER_METRIC_KIND_HISTOGRAM and returns
+/// TRITONSERVER_ERROR_UNSUPPORTED for unsupported TRITONSERVER_MetricKind.
+///
+/// \param metric The metric object to update.
+/// \param value The amount to observe metric's value to.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONSERVER_DECLSPEC struct TRITONSERVER_Error* TRITONSERVER_MetricObserve(
```

Comment:
Reuse `TRITONSERVER_MetricSet`?

Reply:
@rmccorm4 Any thought? Basically we need to merge two funcions below into one `Metric::Set(double value)`. It works but may add confusion.
https://github.com/triton-inference-server/core/blob/fd5c44bae3fb4eaef8b0ac277d68142a8eb11e03/src/metric_family.cc#L338-L404

Reply:
I can take a closer look tomorrow, but my gut reaction is that Guan is probably right and we can probably just reuse Get and Set which will call Collect and Observe internally when `kind == KIND_HISTOGRAM` if functionally equivalent

Reply:
`MetricValue` may not work if `Collect` returns multiple values (one per bucket?), but again will need to take a closer look. Let me know if you already know more details on this from your research.

Reply:
`MetricValue` cannot be reused for histogram.

Reply:
Updated

Reply:
I would like to revisit this change. The consensus is to keep C API and python_backend API 1:1 matched. I am inclined to add a new C API `TRITONSERVER_MetricObserve` for histogram instead of reusing `TRITONSERVER_MetricSet` for three reasons.
1. Both `Histogram` and `Summary` types call `Observe` to record new value. We can reuse observe for `Summary` type if we add it in the future.
2. Histogram also has `ObserveMultiple` API which may be added in the future. I don't like the idea that `Histogram.Set` and `Histogram.ObserveMultiple` coexist.
3. Setting histogram to a value aka Histogram.set(val) is semantically wrong. It is confusing to users familiar with Prometheus APIs. The description of `TRITONSERVER_MetricSet` can be verbose as well in order to describe different behaviors for counter/gauge and histogram/summary.
cc @Tabrizian @rmccorm4 @GuanLuo

Reply:
FYI
https://github.com/jupp0r/prometheus-cpp/blob/master/core/include/prometheus/histogram.h
https://github.com/jupp0r/prometheus-cpp/blob/master/core/include/prometheus/summary.h

Reply:
Well.. Triton metrics API is not supposed to be mirroring "Prometheus API", Prometheus is one of the "forms" we can exhibit the metrics as. So we should design the API to be using generic terms for statistics, the meaning of `gauge`/`counter`/`histogram` (and `summary`?) is not affected by the fact that Prometheus or other metrics libraries are used.

Thinking from this mindset, my question is if `observe` is the generic verb for recording the statistics of `histogram`. If that is the case, then I am fine to add `XXXObserve`, otherwise, we should use the proper verb

Reply:
I think either sample or observe. Voting for Observe for simplicity.
> A histogram samples observations (usually things like request durations or response sizes) and counts them in configurable buckets. It also provides a sum of all observed values.
> 
> Similar to a histogram, a summary samples observations (usually things like request durations and response sizes). While it also provides a total count of observations and a sum of all observed values, it calculates configurable quantiles over a sliding time window.